core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala

```diff
@@ -81,8 +86,10 @@ class ZooKeeperClient(connectString: String, sessionTimeoutMs: Int, connectionTi
       val responseQueue = new ArrayBlockingQueue[Req#Response](requests.size)
 
       requests.foreach { request =>
+        inFlightRequests.acquire()
         send(request) { response =>
           responseQueue.add(response)
+          inFlightRequests.release()
           countDownLatch.countDown()
```

Comment:
incase you missed my previous comment.. just want to remind that we will miss releasing semaphore permits incase of validation errors form zookeeper client. I agree that, this may not be a problem,  if we have enough validation in KafkaZkClient. sample test case given below. 

```java
  @Test
  def testSemaphorePermitLeak(): Unit = {
    import scala.collection.JavaConverters._
    val data = bytes
    var zooKeeperClient: ZooKeeperClient = null
    var createResponse: CreateResponse = null

    try {
      zooKeeperClient = new ZooKeeperClient(zkConnect, zkSessionTimeout, zkConnectionTimeout, 1, null)
      createResponse = zooKeeperClient.handleRequest(CreateRequest("illegalpath", data, ZooDefs.Ids.OPEN_ACL_UNSAFE.asScala,
        CreateMode.PERSISTENT))
    } catch {
      case e: Throwable => info("Exception occured: " + e.getMessage)
    }
    assertEquals("Permit should be 1 ", 1, zooKeeperClient.inFlightRequests.availablePermits())
  }
```

Reply:
Hi @omkreddy. I did see your earlier comment.

While I knew that you should use try-finally with these concurrency primitives to make sure you don't get unexpectedly stuck, I had trouble finding where the problem would actually occur.

Your follow-up comment helped though was still not totally specific. More specifically, `org.apache.zookeeper.ZooKeeper` async apis internally call `org.apache.zookeeper.common.PathUtils.validatePath`, which throws `IllegalArgumentException` if the path is invalid based on some local validation checks.

In which case, I think the change would be the following:
```
requests.foreach { request =>
  inFlightRequests.acquire()
  try {
    send(request) { response =>
      responseQueue.add(response)
      inFlightRequests.release()
      countDownLatch.countDown()
    }
  } finally {
    inFlightRequests.release()
  }
}
```
Is this the scenario you had in mind? Are there others? I agree that this change is needed.

Things get more complicated if we don't trust `org.apache.zookeeper.ZooKeeper` to always deliver responses back, suggesting that the semaphore and countdown latch would never get updated. I'm not sure if we need to address that here.

Reply:
In the above snippet, inFlightRequests.release() is called twice.

Reply:
Good catch @tedyu.

Reply:
@onurkaraman Yes, this is the scenario, i was thinking. 
I agree that your second scenario is bit more complex to implement.  we may need to add zk response timeouts.  If needed , we can fix in the future.

Reply:
Ah. The failing tests reminded me of why this was more complicated than what it initially seemed. We actually want to update both the `inFlightRequests.release()` and `countDownLatch.countDown()` in the callback in the normal case, not the calling thread.

We only want to `inFlightRequests.release()` if the calling thread experiences an exception, in which case we won't even need to `countDownLatch.countDown()` since the calling thread wouldn't have even begun to wait on the countdown latch.